python/python/bystro/covariance/_base_covariance.py

```diff
@@ -0,0 +1,549 @@
+
+    X : np.array-like,(N,sum(idxs))
+        The data
+
+    idxs: np.array-like,(sum(p),)
+        The observation locations
+
+    weights : np.array-like,(N,),default=None
+        The optional weights on the samples
+
+    Returns
+    -------
+    avg_score : float
+        Average log likelihood
+    """
+    if weights is None:
+        weights = np.ones(X.shape[0])
+    avg_score = np.mean(weights * _marginal_score_samples(covariance, X, idxs))
+    return avg_score
+
+
+def _marginal_score_samples(covariance, X, idxs):
+    """
+    Returns the marginal log-likelihood of a subset of data
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    X : np.array-like,(N,sum(idxs))
+        The data
+
+    idxs: np.array-like,(sum(p),)
+        The observation locations
+
+    Returns
+    -------
+    scores : float
+        Average log likelihood
+    """
+    cov1 = covariance[idxs == 1]
+    cov_sub = cov1[:, idxs == 1]
+    scores = _score_samples(cov_sub, X)
+    return scores
+
+
+def _score(covariance, X, weights=None):
+    """
+    Returns the average log liklihood of data.
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    X : np.array-like,(N,sum(p))
+        The data
+
+    weights : np.array-like,(N,),default=None
+        The optional weights on the samples
+
+    Returns
+    -------
+    avg_score : float
+        Average log likelihood
+    """
+    if weights is None:
+        weights = np.ones(X.shape[0])
+    avg_score = np.mean(weights * _score_samples(covariance, X))
+    return avg_score
+
+
+def _score_samples(covariance, X):
+    """
+    Return the log likelihood of each sample
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    X : np.array-like,(N,sum(p))
+        The data
+
+    Returns
+    -------
+    scores : float
+        Log likelihood for each sample
+    """
+    p = covariance.shape[0]
+    term1 = -p / 2 * np.log(2 * np.pi)
+
+    _, logdet = la.slogdet(covariance)
+    term2 = -0.5 * logdet
+
+    quad_init = la.solve(covariance, np.transpose(X))
+    difference = X * np.transpose(quad_init)
+    term3 = np.sum(difference, axis=1)
+
+    scores = term1 + term2 - 0.5 * term3
+    return scores
+
+
+#####################################
+#####################################
+##                                 ##
+##  Information-theoretic methods  ##
+##                                 ##
+#####################################
+#####################################
+
+
+def _entropy(covariance):
+    """
+    Computes the entropy of a Gaussian distribution parameterized by
+    covariance.
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    Returns
+    -------
+    entropy : float
+        The differential entropy of the distribution
+    """
+    covariance.shape[1]
```

Comment:
this line looks like it's a no-op?  FYI ruff should be able to detect these for you, I'll have to doublecheck why it doesn't seem to be picking this one up